larray/core/constrained.py

```diff
@@ -0,0 +1,533 @@
+from abc import ABCMeta
+from copy import deepcopy
+import warnings
+
+import numpy as np
+
+from typing import TYPE_CHECKING, Type, Any, Dict, Set, List, no_type_check
+
+from larray.core.metadata import Metadata
+from larray.core.axis import AxisCollection
+from larray.core.array import Array, empty
+from larray.core.session import Session
+
+
+class NotLoaded:
+    pass
+
+
+try:
+    import pydantic
+except ImportError:
+    pydantic = None
+
+if pydantic:
+    from pydantic.fields import ModelField
+    from pydantic.class_validators import Validator
+    from pydantic.main import BaseConfig
+
+    # the implementation of the class below is inspired by the 'ConstrainedBytes' class
+    # from the types.py module of the 'pydantic' library
+    class ConstrainedArrayImpl(Array):
+        expected_axes: AxisCollection
+        dtype: np.dtype = np.dtype(float)
+
+        # see https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__
+        @classmethod
+        def __get_validators__(cls):
+            # one or more validators may be yielded which will be called in the
+            # order to validate the input, each validator will receive as an input
+            # the value returned from the previous validator
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, value, field: ModelField):
+            # XXX: not sure we should accept a scalar as value but this allows to not repeat axes
+            # when setting a (default) value to the associated array
+            if not (isinstance(value, Array) or np.isscalar(value)):
+                raise TypeError(f"Expected object of type '{Array.__name__}' or a scalar for "
+                                f"the variable '{field.name}' but got object of type '{type(value).__name__}'")
+
+            # check axes
+            if isinstance(value, Array):
+                error_msg = f"Array '{field.name}' was declared with axes {cls.expected_axes} but got array " \
+                            f"with axes {value.axes}"
+                # check for extra axes
+                extra_axes = value.axes - cls.expected_axes
+                if extra_axes:
+                    raise ValueError(f"{error_msg} (unexpected {extra_axes} "
+                                     f"{'axes' if len(extra_axes) > 1 else 'axis'})")
+                # check compatible axes
+                try:
+                    cls.expected_axes.check_compatible(value.axes)
+                except ValueError as error:
+                    error_msg = str(error).replace("incompatible axes", f"Incompatible axis for array '{field.name}'")
+                    raise ValueError(error_msg)
+
+            # broadcast if needed
+            array = empty(axes=cls.expected_axes, dtype=cls.dtype)
+            array[:] = value
```

Comment:
This extra copy is both a good thing (some people find this more intuitive) and a bad thing (a tad less performant but more importantly could lead to unexpected errors given that the behavior is different than for "normal" sessions which only hold references to arrays). Unsure how to solve this (or if it needs solving at all) but this particular point will at least need to be documented explicitly (we possibily need to clarify the situation for normal Session too). 

Also, when is validate called? In the code like below, would it make an extra copy for each line?
```python
s.arr = zeros(...)
s.arr = s.arr + 1
```
If this is the case, can't we imagine a system where if the attribute was never initialized, it uses the current code path using empty, but if it was, only the check is done but not the copy?

Reply:
Good point.
I'll make several tests to see if I can come with something better than this extra copy.

Reply:
I just pushed a new version of `ConstrainedArrayImpl.validate()`. I still need to create an extra copy in case of missing axes but maybe there is another way in case missing axes to avoid to create an extra copy that I don't see.

Reply:
No, there is no way to add axes without copying: you cannot broadcast "inplace". FWIW, there is one exception to this, it is if one accepts the resulting array to be readonly (which .expand() can do as an option) but we clearly cannot use that option here.